### PR TITLE
Huge refactorings

### PR DESCRIPTION
**Pull request by anton31kah**
- Organized code into packages, at root src.
- Added all types to common/models.py with useful methods for conversions and added those types everywhere.
- Moved the singleton metaclass into a separate file.
- Refactored common/publisher.py
- Added config.ini and config_reader.py to temporarily read emqx address.
- Added a logging_service.py empty file, and a metrics empty package.
- Removed unnecessary endpoints in main.py fast-api.
- Heavily refactored src/commands/command_runner.py:
    - Added a method to clean strings, and set whether they should be part of the output. For each line it is checked against it and only newlines are stripped, whitespace is kept, unless the whole line is whitespace. If the line is empty of only whitespace, it won't be added to the output. If the line is multiple lines, then it is split, and the same previous operation is called on each line, then the result is joined.
    - The fields are now private and some expose getter properties.
    - Status is now visible (4 states)
    - Refactored some common code for finishing both kill and run.
    - Added a finishing process for a killed command (return code is stored, status is updated, output is set, and on_exit is called).
    - Added __str__ (toString).
    - Updated all docstrings to match changes.
- Removed unnecessary comments left from copying from multiple tutorials, docs, and guides.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,10 +1,8 @@
-# Default ignored files
-/shelf/
-/workspace.xml
-# Datasource local storage ignored files
-/dataSources/
-/dataSources.local.xml
-# Editor-based HTTP Client requests
-/httpRequests/
```
Comment (NHristovski) [inline .gitignore:8]: Maybe we should leave these because the folder structure is different on other machines

File: .idea/client_app.iml [hunk 2]
```diff
@@ -4,7 +4,10 @@
     <content url="file://$MODULE_DIR$">
```
Comment (NHristovski) [inline .idea/client_app.iml:4]: I am not sure we should change .idea files. We should commit only application specific files

File: .idea/workspace.xml [hunk 3]
```diff
@@ -1,130 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (NHristovski) [inline .idea/workspace.xml:1]: why is this deleted?

File: src/commands/command_handler.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    def run_command(self, command: CommandMessageJSON):
+        command_id = command.command_id
+
+        if command.command_type == CommandType.start:
+            runner = CommandRunner(command_id, command.body, handle_line, handle_result)
+            runner.run()
+            self.active_commands[command_id] = runner
+        elif command.command_type == CommandType.stop:
+            runner = self.active_commands[command_id]
```
Comment (NHristovski) [inline src/commands/command_handler.py]: I think we should add if runner is not null  then runner kill, else throw error / log error.

File: src/commands/command_runner.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    @property
+    def output(self):
+        return self.__output
+
+    @property
+    def is_complete(self) -> bool:
+        """ :return: whether the process completed (both successfully or killed) """
+
+        return self.__status == CommandStatus.Killed or self.__status == CommandStatus.FinishedSuccessfully
```
Comment (NHristovski) [inline src/commands/command_runner.py:113]: i like these 👍

File: src/commands/command_runner.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        return thread
+
+    def kill(self):
+        """ Kills the process and each subprocess launched by the process
+        Also sets status to Killed
+        Still fills output with output of command and calls on_exit
+        """
+
+        if self.__status.Running and self.__process_id:
+            ps_util_process = psutil.Process(self.__process_id)
```
Comment (NHristovski) [inline src/commands/command_runner.py:165]: We should check if ps_util_process is null, if its null as a result we should sent result code 100+ ( the os result codes are from 0 to 20~something i think we can send 100+ result codes as application specific error. For example code 104: You are tying to kill process that is not running

File: src/common/config.ini [hunk 7]
```diff
@@ -0,0 +1,5 @@
+[NETWORK]
```
Comment (NHristovski) [inline src/common/config.ini:1]: masina 💯

File: src/common/singleton.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class Singleton(type):
```
Comment (NHristovski) [inline src/common/singleton.py:2]: I don't think this is thread safe singelton. We should do more research on singleton in python, but this should not be your task. We will have a look at it when we work together.

File: src/main.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+
+""" uvicorn src.main:app --reload """
+
+app = FastAPI()
+
+command_listener = CommandListener("nikola")
+
+
+@app.get("/")
```
Comment (NHristovski) [inline src/main.py:17]: I think we should make this /health. In this method we should check if we are still connected to the mqtt server and return json result  { "status" :  "{if all the services are ok then "ok", else "failed"} ( i think we will have only 1 service so this will be the result of the mqtt service )" , "services" : ["mqtt" : "ok (connected) / "failed" (not connected) ]

**Review by NHristovski — CHANGES_REQUESTED** [review]
Left some comments, not all are for Anton. We can merge this as it is and look at everything when we work together

**Review by voirtimid — APPROVED** [review]
Good job Anton

File: .gitignore (hunk 1 above)
Comment (anton31kah) [reply .gitignore:8]: Solved in https://github.com/NHristovski/ServerHelper-ClientApp/pull/1/commits/20c1f00428f8edfa8feef7a5b8d3a11edae87789.
However, if `.gitignore` is not in the top directory, then it won't work anyway.

File: .idea/client_app.iml (hunk 2 above)
Comment (anton31kah) [reply .idea/client_app.iml:4]: Here I connected `requirements.txt` with PyCharm so it can track changes and update when needed.

File: .idea/workspace.xml (hunk 3 above)
Comment (anton31kah) [reply .idea/workspace.xml:1]: The file shouldn't be uploaded, according to [the official .gitignore by JetBrains](https://github.com/github/gitignore/blob/master/Global/JetBrains.gitignore)

File: src/commands/command_handler.py (hunk 4 above)
Comment (anton31kah) [reply src/commands/command_handler.py]: Resolved in https://github.com/NHristovski/ServerHelper-ClientApp/pull/1/commits/39e1f73cc3ef23743ae2c747feb0de953a4e8f17.

File: src/commands/command_runner.py (hunk 5 above)
Comment (anton31kah) [reply src/commands/command_runner.py:113]: They'll even throw errors if someone from outside the class tries to change them as they are getters only.

File: src/commands/command_runner.py (hunk 6 above)
Comment (anton31kah) [reply src/commands/command_runner.py:165]: Correct, I just checked, and [psutil raises `NoSuchProcess`](https://psutil.readthedocs.io/en/latest/#process-class) if the pid doesn't exist.

Comment (anton31kah) [reply src/commands/command_runner.py:165]: I added the `TODO`s so we can fix this together here https://github.com/NHristovski/ServerHelper-ClientApp/pull/1/commits/fff65d099efc0e2086ce0997f1cf1f1186016596.

File: src/common/singleton.py (hunk 8 above)
Comment (anton31kah) [reply src/common/singleton.py:2]: We can start from [here](https://stackoverflow.com/questions/50566934/). Added a TODO https://github.com/NHristovski/ServerHelper-ClientApp/pull/1/commits/f39bf309121b450fd915a5ec2c0f7da15bc852eb.

File: src/main.py (hunk 9 above)
Comment (anton31kah) [reply src/main.py:17]: Added as a TODO https://github.com/NHristovski/ServerHelper-ClientApp/pull/1/commits/39775126e7794bb268cfb313ad02b6bb0f081cca.

**Comment by anton31kah** [comment]
Sorry for basically adding TODO's for half the stuff but those fixes shouldn't be part of this pull request, even less, my job alone.

**Review by NHristovski — APPROVED** [review]
Great job
